### PR TITLE
Updated Async router with added interface

### PR DESCRIPTION
**Pull request by guyradford**

### CODE REVIEW

File: src/Plugin/Router/AsyncSwitchMessageRouter.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * @package Prooph\ServiceBus\Router
+ * @author Guy Radford <guyr@crazylime.co.uk>
+ */
+class AsyncSwitchMessageRouter implements MessageBusRouterPlugin, ActionEventListenerAggregate
+{
+    use DetachAggregateHandlers;
+
+    /**
+     * @var ActionEventListenerAggregate
```
Comment (codeliner) [inline src/Plugin/Router/AsyncSwitchMessageRouter.php]: type hint should be the new `MessageBusRouterPlugin` interface

File: src/Plugin/Router/AsyncSwitchMessageRouter.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * @var MessageProducer
+     */
+    protected $asyncMessageProducer;
+
+
+    /**
+     * @param ActionEventListenerAggregate $router
+     * @param MessageProducer $asyncMessageProducer
+     */
+    public function __construct(ActionEventListenerAggregate $router, MessageProducer $asyncMessageProducer)
```
Comment (codeliner) [inline src/Plugin/Router/AsyncSwitchMessageRouter.php]: type hint should be the new `MessageBusRouterPlugin` interface

File: docs/plugins.md [hunk 3]
```diff
@@ -83,6 +83,31 @@ $router->route('/^My\.Event\.Article.*/')->to(new InventoryUpdater());
 $eventBus->utilize($router);
 ```
 
+### Prooph\ServiceBus\Plugin\Router\AsyncSwitchMessageRouter
+
+The `AsyncSwitchMessageRouter` Router allows you to easily set up a single router to handle both your async and sync messages.
+
+To send messages via the Async Provider mark them with the `Prooph\ServiceBus\Async\AsyncMessage` interface.
```
Comment (codeliner) [inline docs/plugins.md]: s/Provider/Producer

File: docs/plugins.md [hunk 4]
```diff
@@ -83,6 +83,31 @@ $router->route('/^My\.Event\.Article.*/')->to(new InventoryUpdater());
...
+
+The `AsyncSwitchMessageRouter` is a decorator that wraps your router. The first time the `AsyncSwitchMessageRouter` sees an async message it is sent via the Async Provider, after that the message is routed normally.
```
Comment (codeliner) [inline docs/plugins.md]: s/Provider/Producer

**Comment by codeliner** [comment]
This PR supersedes #130

**Comment by guyradford** [comment]
@codeliner 

Good morning,

I am not total clear where the **s/Provider/Producer** goes.... is it to replace **Provider** in ..
```
...sent via the Async Provider, after that the message...
```
like so

```
...sent via the Async Provider/Producer, after that the message...
```

then I do understand the **s/** at the start?

Thank you

**Comment by guyradford** [comment]
@codeliner Should `Prooph\ServiceBus\Plugin\Guard\RouteGuard` be using the `MessageBusRouterPlugin` interface?

**Comment by prolic** [comment]
@guyradford 

...sent via the Async Provider, after that the message...

should be

...sent via the Async Producer, after that the message...

File: docs/plugins.md [hunk 5]
```diff
@@ -83,6 +83,31 @@ $router->route('/^My\.Event\.Article.*/')->to(new InventoryUpdater());
...
+To send messages via the Async Producer mark them with the `Prooph\ServiceBus\Async\AsyncMessage` interface.
+
+The `AsyncSwitchMessageRouter` is a decorator that wraps your router. The first time the `AsyncSwitchMessageRouter` sees an async message it is sent via the Async Producer, after that the message is routed normally.
+
+```
+//You can define your primary router...
+$myRouter = new MyRouter();
+
+// Create async message producer...
+$asyncMessageProducer = new AsyncMessageProducerFactory();
```
Comment (prolic) [inline docs/plugins.md]: remove `Factory`

File: tests/Plugin/Router/AsyncSwitchMessageRouterTest.php [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        $rtn = $router->onRouteMessage($actionEvent);
+
+        $this->assertEquals('handled-by-decorated-router', $rtn);
+        $updatedMessage = $actionEvent->getParam(MessageBus::EVENT_PARAM_MESSAGE);
+        $this->assertArrayHasKey('handled-async', $updatedMessage->metadata());
+        $this->assertTrue($updatedMessage->metadata()['handled-async']);
+    }
+}
+
+class AsyncCommand extends Command implements PayloadConstructable, AsyncMessage
```
Comment (prolic) [inline tests/Plugin/Router/AsyncSwitchMessageRouterTest.php]: Move that class to Mock namespace

File: tests/Plugin/Router/AsyncSwitchMessageRouterTest.php [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+     * @return AsyncCommand
+     */
+    public static function createCommand($data)
+    {
+        return new self([
+            'data' => $data
+        ]);
+    }
+}
+class NonAsyncCommand extends Command implements PayloadConstructable
```
Comment (prolic) [inline tests/Plugin/Router/AsyncSwitchMessageRouterTest.php]: Move that class to Mock namespace

File: src/Plugin/Router/AsyncSwitchMessageRouter.php (hunk 1 above)
Comment (prolic) [inline src/Plugin/Router/AsyncSwitchMessageRouter.php]: still this

**Comment by codeliner** [comment]
@guyradford 
> @codeliner Should Prooph\ServiceBus\Plugin\Guard\RouteGuard be using the MessageBusRouterPlugin interface?

Good question, but I would say no, because the `RouteGuard` is not a router plugin. It just listens on the same "action event" like a router does.

**Comment by prolic** [comment]
Exactly

Am 29.08.2016 2:37 vorm. schrieb "Alexander Miertsch" <
notifications@github.com>:

**Comment by guyradford** [comment]
ok, great :) assuming there is no more feedback, I think I am done.

Thank you for your help :)

**Comment by codeliner** [comment]
@guyradford you've missed a feedback ;) so sorry, but still a small thing to do, see here: https://github.com/prooph/service-bus/pull/132/files#r76523161

Use the correct type hint and the PR is ready. Thank you again for the great work

File: src/Plugin/Router/AsyncSwitchMessageRouter.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+     * @var MessageBusRouterPlugin
+     */
+    protected $asyncMessageProducer;
+
+
+    /**
+     * @param ActionEventListenerAggregate $router
+     * @param MessageBusRouterPlugin $asyncMessageProducer
+     */
+    public function __construct(ActionEventListenerAggregate $router, MessageBusRouterPlugin $asyncMessageProducer)
```
Comment (codeliner) [inline src/Plugin/Router/AsyncSwitchMessageRouter.php]: same as above, wrong type hint

**Comment by codeliner** [comment]
:fireworks:

**Comment by guyradford** [comment]
Thank you for all you help :)

**Pull request merged**